File: MapsF/lib/models/field_observation.dart
class FieldObservation {
  const FieldObservation({
    required this.id,
    required this.mapProjectId,
    required this.userId,
    required this.title,
    required this.lat,
    required this.lng,
    required this.createdAt,
    this.description,
    this.observationType,
    this.accuracy,
    this.photoPath,
    this.properties,
  });

  final String id;
  final String mapProjectId;
  final String userId;
  final String title;
  final String? description;
  final String? observationType;
  final double lat;
  final double lng;
  final double? accuracy;
  final String? photoPath;
  final Map<String, dynamic>? properties;
  final DateTime createdAt;

  factory FieldObservation.fromJson(Map<String, dynamic> json) {
    return FieldObservation(
      id: json['id'].toString(),
      mapProjectId: json['map_project_id'].toString(),
      userId: json['user_id'].toString(),
      title: json['title'].toString(),
      description: json['description']?.toString(),
      observationType: json['observation_type']?.toString(),
      lat: double.parse(json['lat'].toString()),
      lng: double.parse(json['lng'].toString()),
      accuracy: double.tryParse(json['accuracy']?.toString() ?? ''),
      photoPath: json['photo_path']?.toString(),
      properties: json['properties'] is Map<String, dynamic>
          ? json['properties'] as Map<String, dynamic>
          : null,
      createdAt:
          DateTime.tryParse(json['created_at']?.toString() ?? '') ??
          DateTime.now(),
    );
  }
}

File: MapsG/app/models/field_observation.py
from datetime import UTC, datetime

from sqlalchemy import DateTime, Float, ForeignKey, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.core.ids import new_id


class FieldObservation(Base):
    __tablename__ = "field_observations"

    id: Mapped[str] = mapped_column(String(64), primary_key=True, default=lambda: new_id("observation"))
    map_project_id: Mapped[str] = mapped_column(String(64), ForeignKey("map_projects.id"), index=True)
    user_id: Mapped[str] = mapped_column(String(64), ForeignKey("users.id"), index=True)
    title: Mapped[str] = mapped_column(String(200))
    description: Mapped[str | None] = mapped_column(Text)
    observation_type: Mapped[str | None] = mapped_column(String(100))
    lat: Mapped[float] = mapped_column(Float)
    lng: Mapped[float] = mapped_column(Float)
    accuracy: Mapped[float | None] = mapped_column(Float)
    photo_path: Mapped[str | None] = mapped_column(Text)
    properties: Mapped[dict | None] = mapped_column(JSON)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    synced_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

